File: leetcode/algorithms/700_search_in_a_binary_search_tree/main.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class SearchInABinarySearchTree:
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if root != None:
            if root.val == val:
                return root

            left: Optional[TreeNode] = self.searchBST(root=root.left, val=val)
            if left != None:
                return left

            right: Optional[TreeNode] = self.searchBST(
                root=root.right, val=val)
            if right != None:
                return right

        return None


    # Solution
    # Solution 1: Recursion
    def solution1(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if root == None:
            return root

        if root.val == val:
            return root

        return self.solution1(root=root.left, val=val) if root.val > val else self.solution1(root=root.right, val=val)

    # Solution 2: Iteration
    def solution2(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        while root != None and root.val != val:
            root = self.solution2(root=root.left, val=val) if root.val > val else self.solution2(root=root.right, val=val)
        return root

File: leetcode/algorithms/700_search_in_a_binary_search_tree/main.java
class SearchInABinarySearchTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public TreeNode searchBST(TreeNode root, int val) {
        if (root != null) {
            if (root.val == val) {
                return root;
            }

            TreeNode left = searchBST(root.left, val);
            if (left != null) {
                return left;
            }

            TreeNode right = searchBST(root.right, val);
            if (right != null) {
                return right;
            }
        }

        return null;
    }


    // Solution
    // Solution 1: Recursion
    public TreeNode solution1(TreeNode root, int val) {
        if (root == null) {
            return root;
        }

        if (root.val == val) {
            return root;
        }

        return root.val > val ? solution1(root.left, val) : solution1(root.right, val);
    }

    // Solution 2: Iteration
    public TreeNode solution2(TreeNode root, int val) {
        while (root != null && root.val != val) {
            root = root.val > val ? solution2(root.left, val) : solution2(root.right, val);
        }
        return root;
    }
}
